Type warnings

BridgeAR opened the pull request:
This is on top of #3836 and I wanted to get your feedback what you think about the implementation. So you only have to look at the "Initial commit" changes.

If someone uses a data type in a way that is not supported we'd now warn the user about it including what we do instead and presenting the link to the data types as they are in the current dialect.

janmeier commented:
I like this implementation very much! :) LGTM

BridgeAR commented:
This will close #2038

mickhansen commented on lib/dialects/mssql/data-types.js:
```diff
@@ -68,13 +76,85 @@ DATE.prototype.toSql = function() {
+
+  // MSSQL does not support any parameters for integer
+  if (this._length || this.options.length || this._unsigned || this._zerofill) {
+    this.warn('You used INTEGER with a attribute. This is not supported by MSSQL. Simple `INTEGER` will be used instead.');
```
'an attribute'

mickhansen commented on lib/data-types.js:
```diff
@@ -38,12 +39,20 @@ var ABSTRACT = function(options) {
+ABSTRACT.prototype.warn = function(text) {
+  if (!warnings[text]) {
+    warnings[text] = true;
+    console.warn('>> WARNING:', text, '\n>> Check:', this.dialectTypes);
```
Hmm, any way to have this include the actual line ala what depd does you think?

janmeier replied:
Couldn't we actually just use depd here?

mickhansen replied:
It's not really a deprecation, i'm not sure if depd explicitely uses that wording though to.

BridgeAR replied:
Hm that would be nice indeed... It seems like depd [includes 'deprecated' hardcoded](https://github.com/dougwilson/nodejs-depd/blob/master/index.js#L100). I could include the colours without difficulties though :D

mickhansen replied:
Could duplicate the stack functionality though.

BridgeAR replied:
I thought about it but let's move that to another day and another PR :)

mickhansen commented on lib/dialects/postgres/data-types.js:
```diff
@@ -55,10 +65,13 @@ var INTEGER = function() {
-  this._unsigned = null;
-  this._zerofill = null;
+  if (this._length || this.options.length || this._unsigned || this._zerofill) {
+    this.warn('You used INTEGER with a parameter. This is not supported by POSTGRES. Simple `INTEGER` will be used instead.');
```
Here it's 'parameter' the other places it's 'attribute', intentional?

BridgeAR replied:
I'm unhappy with the wording but nothing better came to my mind. The difference I tried to emphasize is between `BLOB('tiny')` and `INTEGER.UNSIGNED`.

mickhansen replied:
I see, that difference isn't really obvious at all (to me atleast), so i'd say just use "parameter" for all.

BridgeAR replied:
Done

mickhansen commented:
Looks great. Warnings are the way to go, since i don't personally thing we should ever remove support completely.

BridgeAR commented:
Rebased, fixed a couple typos and included more warnings. Depending on what you wish I'd say it's ready to merge. We could provide the additional functionality to include the lines later on.

mickhansen commented:
You're still mixing `attribute` and `parameter` in the warnings, would be nice to streamline them. Or provide some reasoning why they are different.

mickhansen commented:
General wording be better with something like 'postgres does not support TEXT with size. Simple 'TEXT' will be used instead`

BridgeAR commented:
+1

BridgeAR commented:
Changed the comments and rebased

diosney commented:
Please, see https://github.com/sequelize/sequelize/issues/6385

I really don't like this PR, there is a way to disable completely this messages?
